### PR TITLE
FileTarget: Fix compress archive file not working when using concurrentWrites="True" and keepFileOpen="True"

### LINKED ISSUE #1028: File target with  concurrentWrites="True" and enableArchiveFileCompression="True" does not function
Here is my File target specification:

     <target name="FileLogger" xsi:type="File" header = "------ Connexion v${assembly-version} ------" layout="${longdate} (${threadid}) ${level} ${event-context:eventId} [${logger}] ${share.TruncateLayout:${message}} ${exception:format=tostring}" fileName="${specialFolder:folder=CommonApplicationData}/ConnexionV14/logs/Connexion.txt" archiveFileName="${specialFolder:folder=CommonApplicationData}/ConnexionV14/logs/Connexion.{#}.zip" archiveNumbering="Date" archiveEvery="Day" archiveDateFormat="yyyyMMdd" maxArchiveFiles="10" enableArchiveFileCompression="true" concurrentWrites="true" keepFileOpen="true" encoding="iso-8859-2" />

The offending line in void ArchiveFile(string fileName, string archiveFileName, bool enableCompression) appears to here

                using (var originalFileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read))

An exception occurs because another file is using it.

If I change this to

               using (var originalFileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete))
 
Which matches the original sharing specification when the file was opened. My understanding, and I know I have found documentation on this before, is that the FileShare specification needs to match the currently opened file exactly, or the share will fail.

At any rate, the exception is thrown and I am left with a bunch of 0  length zip files. The main log file never rotates.

If I remove the enableArchiveFileCompression flag, everything works, but there is no compression.

Also, do you think it would be better to perform the compression after the file has been rotated? This could be done in a background thread.

### PR DESCRIPTION
**Pull request by bhaeussermann**
Fixes #1028

### CODE REVIEW

**Comment by bhaeussermann** [comment]
I couldn't find a way to unit-test this as the log file needs to be locked by an external process :(.

File: src/NLog/Targets/FileTarget.cs [hunk 1]
```diff
@@ -1087,7 +1087,16 @@ private static void ArchiveFile(string fileName, string archiveFileName, bool en
                     }
                 }
 
-                File.Delete(fileName);
+                if (ConcurrentWrites && KeepFileOpen)
+                {
+                    // Moving the file will cause all other FileTarget's to release the file, unlocking it for deleting.
+                    string tempFileName = Path.GetTempFileName();
```
Comment (304NotModified) [inline src/NLog/Targets/FileTarget.cs]: I think this is a hack?

**Comment by bhaeussermann** [comment]
> I couldn't find a way to unit-test this as the log file needs to be locked by an external process :(.

It turns out the other `FileTarget` need not run in a separate process for the file locking to occur :). I supplemented the existing `DateArchive_AllLoggersTransferToCurrentLogFile()` test with a new `enableArchiveCompression` parameter and confirmed that it exposes the bug.

**Comment by 304NotModified** [comment]
Nice that you could reproduce it with an unit test :)

**Comment by 304NotModified** [comment]
Thank! It has been merged :)

**Pull request merged**